include/master_element/MasterElementFunctions.h

```diff
@@ -28,82 +28,82 @@
 namespace sierra {
 namespace nalu {
 
+  template<typename ftype> inline void cofactorMatrix(ftype adjJac[][3], const ftype jact[][3]) {
+    adjJac[0][0] = jact[1][1] * jact[2][2] - jact[2][1] * jact[1][2];
+    adjJac[0][1] = jact[1][2] * jact[2][0] - jact[2][2] * jact[1][0];
+    adjJac[0][2] = jact[1][0] * jact[2][1] - jact[2][0] * jact[1][1];
+
+    adjJac[1][0] = jact[0][2] * jact[2][1] - jact[2][2] * jact[0][1];
+    adjJac[1][1] = jact[0][0] * jact[2][2] - jact[2][0] * jact[0][2];
+    adjJac[1][2] = jact[0][1] * jact[2][0] - jact[2][1] * jact[0][0];
+
+    adjJac[2][0] = jact[0][1] * jact[1][2] - jact[1][1] * jact[0][2];
+    adjJac[2][1] = jact[0][2] * jact[1][0] - jact[1][2] * jact[0][0];
+    adjJac[2][2] = jact[0][0] * jact[1][1] - jact[1][0] * jact[0][1];
+  }
+  template<typename ftype> inline void cofactorMatrix(ftype adjJac[][2], const ftype jact[][2]) {
+    adjJac[0][0] =  jact[1][1];
+    adjJac[0][1] = -jact[1][0];
+    adjJac[1][0] = -jact[0][1];
+    adjJac[1][1] =  jact[0][0];
+  }
+
   template <typename AlgTraits, typename GradViewType, typename CoordViewType, typename OutputViewType>
-  void generic_grad_op_3d(const GradViewType& referenceGradWeights, const CoordViewType& coords, OutputViewType& weights)
+  void generic_grad_op(const GradViewType& referenceGradWeights, const CoordViewType& coords, OutputViewType& weights)
   {
+    constexpr int dim = AlgTraits::nDim_;
+
     using ftype = typename CoordViewType::value_type;
     static_assert(std::is_same<ftype, typename GradViewType::value_type>::value,  "Incompatiable value type for views");
     static_assert(std::is_same<ftype, typename OutputViewType::value_type>::value,  "Incompatiable value type for views");
-    static_assert(GradViewType::Rank == 3, "grad view assumed to be 3D");
-    static_assert(CoordViewType::Rank == 2, "Coordinate view assumed to be 2D");
-    static_assert(OutputViewType::Rank == 3, "Weight view assumed to be 3D");
-    static_assert(AlgTraits::nDim_ == 3, "3D method");
+    static_assert(GradViewType::Rank   ==   3, "grad view assumed to be 3D");
+    static_assert(CoordViewType::Rank  ==   2, "Coordinate view assumed to be 2D");
+    static_assert(OutputViewType::Rank ==   3, "Weight view assumed to be 3D");
```

Comment:
A bunch of rank asserts.  Would probably not compile if the wrong rank object were passed anyway but this will give a clearer compile error instead of something like an incompatible data types error in the body of the function.